cluster-operator/src/main/java/io/strimzi/operator/cluster/PlatformFeaturesAvailability.java

```diff
@@ -4,23 +4,138 @@
  */
 package io.strimzi.operator.cluster;
 
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.VersionInfo;
 import io.strimzi.operator.cluster.operator.KubernetesVersion;
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 /**
  * Gives a info about certain features availability regarding to kubernetes version
  */
 public class PlatformFeaturesAvailability {
+    private static final Logger log = LogManager.getLogger(PlatformFeaturesAvailability.class.getName());
 
-    private boolean isOpenshift;
-    private final KubernetesVersion kubernetesVersion;
+    private boolean routes = false;
+    private boolean builds = false;
+    private boolean images = false;
+    private boolean apps = false;
+    private KubernetesVersion kubernetesVersion;
 
-    public PlatformFeaturesAvailability(boolean isOpenshift, KubernetesVersion kubernetesVersion) {
-        this.isOpenshift = isOpenshift;
+    public static Future<PlatformFeaturesAvailability> create(Vertx vertx, KubernetesClient client) {
+        Future<PlatformFeaturesAvailability> pfaFuture = Future.future();
+        Future<Void> composeFuture = Future.future();
+        OkHttpClient httpClient = getOkHttpClient(client);
+
+        PlatformFeaturesAvailability pfa = new PlatformFeaturesAvailability();
+
+        Future<VersionInfo> futureVersion = getVersionInfo(vertx, client);
+
+        futureVersion.compose(versionInfo -> {
+            String major = versionInfo.getMajor().equals("") ? Integer.toString(KubernetesVersion.MINIMAL_SUPPORTED_MAJOR) : versionInfo.getMajor();
+            String minor = versionInfo.getMinor().equals("") ? Integer.toString(KubernetesVersion.MINIMAL_SUPPORTED_MINOR) : versionInfo.getMinor();
+            pfa.setKubernetesVersion(new KubernetesVersion(Integer.parseInt(major.split("\\D")[0]), Integer.parseInt(minor.split("\\D")[0])));
+
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "route.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setRoutes(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "build.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setBuilds(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "apps.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setApps(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "image.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setImages(supported);
+            composeFuture.complete();
+        }, composeFuture);
+
+        composeFuture.setHandler(res -> {
```

Comment:
Can this not be replaced by `composeFuture.setHandler(pfaFuture.completer())`

Reply:
Yeah, I think this whole part is completely unnecessary. I removed the whole block and return the future from the compose directly.